walletnode/services/artworksearch/thumbnail/thumbnail.go

```diff
@@ -0,0 +1,163 @@
+package thumbnail
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/pastel"
+	nodeClient "github.com/pastelnetwork/gonode/walletnode/node"
+	"github.com/pastelnetwork/gonode/walletnode/services/artworkregister/node"
+)
+
+const (
+	maxConnections    = 10
+	maxConnectionsErr = "cannot connect to more than 10 nodes"
+)
+
+// Helper is interface contract for Thumbnail Getter
+type Helper interface {
+	Connect(ctx context.Context, connections uint) error
+	Fetch(ctx context.Context, key string) ([]byte, error)
+	Close()
+}
+
+type thumbnailHelper struct {
+	pastelClient pastel.Client
+	nodeClient   nodeClient.Client
+
+	reqCh      chan request
+	timeOut    time.Duration
+	isClosed   bool
+	closeMutex sync.Mutex
+}
+
+// New returns a new instance of thumbnailHelper as Helper
+func New(pastelClient pastel.Client, nodeClient nodeClient.Client, timeout time.Duration) Helper {
+	return &thumbnailHelper{
+		pastelClient: pastelClient,
+		nodeClient:   nodeClient,
+		timeOut:      timeout,
+		reqCh:        make(chan request),
+	}
+}
+
+type response struct {
+	data []byte
+	err  error
+}
+
+type request struct {
+	key    string
+	respCh chan *response
+}
+
+// Connect creates `conncetions` no. of connections with supernode & starts thumbnail request listeners
+func (t *thumbnailHelper) Connect(ctx context.Context, connections uint) error {
+	if connections > maxConnections {
+		return errors.New(maxConnectionsErr)
+	}
+
+	nodes, err := t.pastelTopNodes(ctx, connections)
+	if err != nil {
+		return fmt.Errorf("pastelTopNodes: %v", err)
+	}
+
+	for _, node := range nodes {
+		node := node
+		if err := node.Connect(ctx, t.timeOut); err != nil {
+			return err
```

Comment:
you return if any connect error - but leave some previous connect success without closing it.
I think we can consider two options :
1. if any error - stop all listening goroutine
2. return error if all connect() return error, and return ok if any connect got success.

Reply:
applied (1) I think that makes more sense